### ISSUE TITLE
Order of validations

### ISSUE THREAD

**Issue by codepunkt** [jquense/yup#256]
If have this validation schema that i use to validate a youtube video id as part of a multi page form wizard:

```javascript
const firstPageSchema = yup.object().shape({
  youtubeId: yup
    .string()
    .min(11, 'has to be 11 characters')
    .max(11, 'has to be 11 characters')
    .matches(
      /^[a-zA-Z0-9_-]{11}$/,
      'allowed characters are letters, numbers, underscore and dash'
    )
    .test(
      'youtube-api',
      'video not found or not embeddable',
      async value => {
        const parts = 'snippet,contentDetails,status'
        const response = await fetch(
          `${YOUTUBE_API_V3}/videos?id=${value}&key=${YOUTUBE_API_KEY}&part=${parts}`
        )
        const body = await response.json()
        return body.items.length !== 0 && body.items[0].status.embeddable
      }
    )
    .required(),
})
```

Whenever i validate a string that has less than 11 characters, the `.test` is executed and tries to fetch information about the video from youtube.

I would like the `.test` to only run when `.min`, `.max` and `.matches` validation did not produce an error.
How would i do that?

I tried `when` on the same field, but that seems to be forbidden.

**Comment by jquense** [comment]
Hey, this is not really possible at moment because validation s are run in parallel. There are few issues in the (one this week) covering it. We are open to prs!

**Comment by rynjwssl** [comment]
Have there been any updates with this? Sequential validation would definitely be useful. Also, I'm not sure but it seems like `.sequence` is something that was custom-built for the previous example and not part of the official yup API. Is that correct?

**Comment by evankennedy** [comment]
This would be pretty valuable. For anyone coming here wondering if they can do the same thing, you could save the ID format schema and call it within your other schema.

```js
const youtubeIdSchema = yup.string()
  .length(11, 'has to be 11 characters')
  .matches(
    /^[a-zA-Z0-9_-]{11}$/,
    'allowed characters are letters, numbers, underscore and dash'
  );

const firstPageSchema = yup.object({
  youtubeId: youtubeIdSchema
    .test(
      'youtube-api',
      'video not found or not embeddable',
      async value => {
        if (await youtubeIdSchema.isValid(value)) {
          const parts = 'snippet,contentDetails,status';
          const response = await fetch(
            `${YOUTUBE_API_V3}/videos?id=${value}&key=${YOUTUBE_API_KEY}&part=${parts}`
          );
          const body = await response.json();
          return body.items.length !== 0 && body.items[0].status.embeddable;
        }
        return true;
      }
    )
    .required(),
});
```

That would double-validate the first schema, but those are simple validations so that's probably okay.

**Comment by renanBritz** [comment]
Just my two cents on this...

```javascript
const firstPageSchema = yup.object().shape({
  youtubeId: yup
    .test('youtube-api', 'video not found or not embeddable', async (value) => {
      await yup.object.shape({
        youtubeId: yup.string()
          .required()
          .min(11, 'has to be 11 characters')
          .max(11, 'has to be 11 characters')
          .matches(/^[a-zA-Z0-9_-]{11}$/, 'allowed characters are letters, numbers, underscore and dash');
      }).validate({ youtubeId: value });
          
      const parts = 'snippet,contentDetails,status';
      const response = await fetch(`${YOUTUBE_API_V3}/videos?id=${value}&key=${YOUTUBE_API_KEY}&part=${parts}`);
      const body = await response.json();
      return body.items.length !== 0 && body.items[0].status.embeddable;
    })
  })
```
By doing this, if the string validation fails inside the .test method, it will throw a ValidationError which will prevent the api call from running and also replace the error message from .test

**Comment by illia-chaban** [comment]
@renanBritz answer didn't work for me. 
For anyone else looking for an answer or a slightly nicer interface, here's a reusable function that worked for me (Typescript).
I didn't want to use .addMethod because it would require me to update global.d.ts file 

```ts
const yupTestSequence = <T>(
  name: string,
  ...schemas: ReadonlyArray<yup.BaseSchema>
): yup.TestConfig<T> => {
  return {
    name,
    // Has to be a function. For more ref: https://stackoverflow.com/a/63808896/16538301
    test: async function (value) {
      try {
        for (const schema of schemas) {
          await schema.validate(value)
        }
      } catch (error) {
        if (error instanceof yup.ValidationError) {
          // returning the error doesn't work for some reason (it's just being ignored)
          return this.createError({ message: error.message })
        }
        throw error
      }
      return true
    },
  }
}

// Example

  const schema = yup.object({
    emailAddress: yup
      .string()
      .test(
        yupTestSequence(
          'email',
          yup.string().required().email('Enter a valid email'),
          yup.string().test({
            name: 'duplicate email',
            message: 'This email is already in use. Please enter a different email or log in to continue'
            test: async (email) => !(await getEmailAddressExists( email )),
          }),
        ),
      ),
  })
```

**Comment by Qiming-Liu** [comment]
https://github.com/jquense/yup/issues/851#issuecomment-1049705180
